inference-engine/src/transformations/src/transformations/common_optimizations/shuffle_channels_fusion.cpp

```diff
@@ -0,0 +1,118 @@
+
+    is_transformation_valid &= is_reshape_before_valid & is_transpose_valid & is_reshape_after_valid;
+    return is_transformation_valid;
+}
+
+NGRAPH_RTTI_DEFINITION(ngraph::pass::ShuffleChannelsFusion, "ShuffleChannelsFusion", 0);
+
+ngraph::pass::ShuffleChannelsFusion::ShuffleChannelsFusion(const bool reshape_constants_check) {
+    MATCHER_SCOPE(ShuffleChannelsFusion);
+    auto input = ngraph::pattern::any_input(pattern::has_static_shape());
+    auto reshape_before_const_pattern = ngraph::pattern::wrap_type<ngraph::opset6::Constant>(pattern::has_static_shape());
+    auto transpose_const_pattern = ngraph::pattern::wrap_type<ngraph::opset6::Constant>(pattern::has_static_shape());
+    auto reshape_after_const_pattern = ngraph::pattern::wrap_type<ngraph::opset6::Constant>(pattern::has_static_shape());
+
+    auto has_static_shape_and_single_consumer = [](const Output<Node>& output) {
+        return pattern::has_static_shape() && pattern::consumers_count(1);
+    };
+    auto reshape_before_pattern = ngraph::pattern::wrap_type<ngraph::opset6::Reshape>({input, reshape_before_const_pattern},
+                                                                                      has_static_shape_and_single_consumer);
+    auto transpose_pattern = ngraph::pattern::wrap_type<ngraph::opset6::Transpose>({reshape_before_pattern, transpose_const_pattern},
+                                                                                   has_static_shape_and_single_consumer);
+    auto reshape_after_pattern = ngraph::pattern::wrap_type<ngraph::opset6::Reshape>({transpose_pattern, reshape_after_const_pattern},
+                                                                                     pattern::has_static_shape());
+
+    ngraph::matcher_pass_callback callback = [=](pattern::Matcher& m) {
+        const auto& pattern_map = m.get_pattern_value_map();
+
+        auto data = pattern_map.at(input);
+        auto reshape_before = std::dynamic_pointer_cast<ngraph::opset6::Reshape>(pattern_map.at(reshape_before_pattern).get_node_shared_ptr());
+        auto transpose = std::dynamic_pointer_cast<ngraph::opset6::Transpose>(pattern_map.at(transpose_pattern).get_node_shared_ptr());
+        auto reshape_after = std::dynamic_pointer_cast<ngraph::opset6::Reshape>(pattern_map.at(reshape_after_pattern).get_node_shared_ptr());
+        if (!reshape_after || !transpose || !reshape_after)
+            return false;
+
+        if (reshape_constants_check) {
+            auto reshape_before_constant = std::dynamic_pointer_cast<ngraph::opset6::Constant>(
+                pattern_map.at(reshape_before_const_pattern).get_node_shared_ptr());
+            auto reshape_after_constant = std::dynamic_pointer_cast<ngraph::opset6::Constant>(
+                pattern_map.at(reshape_after_const_pattern).get_node_shared_ptr());
+
+            const auto& reshape_before_values = reshape_before_constant->cast_vector<int64_t>();
```

Comment:
You always need to check that the result of dynamic cast is not nullptr.

Reply:
Thanks, done